src/main/webapp/app/exam/participate/exam-participation.component.ts

```diff
@@ -162,6 +177,48 @@ export class ExamParticipationComponent implements OnInit, OnDestroy, ComponentC
                     },
                     error: () => (this.loadingExam = false),
                 });
+            } else if (this.testExam && !this.studentExamId) {
```

Comment:
Please simplify `ngOnInit()`

There are too many special cases and it has too much duplicated code. This is really hard to maintain and leaves plenty of rooms for edge case issues. Ideally all 3 exam types are handled completely the same
1. Test runs (by instructors)
2. Test exams (by students)
3. Real exams

Also on the server side those cases should (mostly) be treated similarly. Otherwise, we always need to completely test them separately, which would lead to a very high testing effort!